sru/ops.py

```diff
@@ -89,25 +89,56 @@ def elementwise_recurrence_gpu(U: Tensor,
                                has_skip_term: bool,
                                scale_x: Optional[Tensor] = None,
                                dropout_mask_c: Optional[Tensor] = None,
-                               mask_pad: Optional[Tensor] = None) -> List[Tensor]:
+                               mask_pad: Optional[Tensor] = None,
+                               amp_recurrence_fp16: bool = False) -> List[Tensor]:
     """Elementwise forward operation of SRU on GPU.
 
     """
     from .cuda_functional import SRU_Compute_GPU
-    return SRU_Compute_GPU.apply(
-        U,
-        x,
-        weight_c,
-        bias,
-        c_init,
-        activation_type,
-        hidden_size,
-        bidirectional,
-        has_skip_term,
-        scale_x,
-        dropout_mask_c,
-        mask_pad
-    )
+
+    in_autocast = getattr(torch, 'is_autocast_enabled', lambda: False)()
+    if in_autocast:
+        with torch.cuda.amp.autocast(enabled=False):
+            cast = torch.Tensor.half if amp_recurrence_fp16 else torch.Tensor.float
+
+            U = cast(U)
+            x = cast(x)
+            weight_c = cast(weight_c)
+            bias = cast(bias)
+            c_init = cast(c_init)
+            scale_x = cast(scale_x) if scale_x is not None else scale_x
+            dropout_mask_c = cast(dropout_mask_c) if dropout_mask_c is not None else dropout_mask_c
+            mask_pad = cast(mask_pad) if mask_pad is not None else mask_pad
```

Comment:
dropout_mask_c should be cast as fp16 or fp32. but mask_pad should remain unchanged as byte / bool tensor

Reply:
Just curious: would `cast(tensor)` do nothing if `tensor` has already the dtype?

Reply:
Right. it becomes a `nop`. eg `float` is a `nop` if already `float32`:
<img width="639" alt="Screen Shot 2020-10-03 at 1 44 52 PM" src="https://user-images.githubusercontent.com/20174596/94998278-a825cd80-057e-11eb-96b9-36579d5eb2d2.png">

(from https://pytorch.org/docs/stable/amp.html )